runtime/doc/map.txt

```diff
@@ -1431,6 +1431,223 @@ Possible values are (second column is the short name used in listing):
     -addr=other		  ?	other kind of range
 
 
+Incremental preview ~
+                                                  *:command-preview* {nvim-api}
+It's possible to allow commands to be incrementally previewed when
+'inccommand' is enabled by using the preview command attribute. The preview
+attribute may only be used through the API. (see |nvim_create_user_command()|)
+
+The value of the preview attribute must be either a boolean value, the name of
+a Vim script function or a |Lua| function callback. Here are some examples: >
+
+	" setting preview to a boolean value
+	call nvim_create_user_command('foo', 'call Foo()',
+		\ { 'preview': v:true })
+
+	" setting preview to a Vim script function
+	function! PreviewFunc()
+	    return v:true
+	endfunction
+
+	call nvim_create_user_command('bar', 'call Bar()',
+		\ { 'preview': 'PreviewFunc' })
+
+	-- setting preview to a Lua function callback
+	lua << EOF
+	local function PreviewFunc()
+	    return true
+	end
+
+	vim.api.nvim_create_user_command('baz', 'call Baz()',
+	                                 { preview = PreviewFunc })
+	EOF
+<
+
+If preview is |TRUE|, preview is always triggered every time the command line
+is modified. If it's not a boolean value, the Vim script or Lua function is
+called and command preview is triggered only if the function returns |TRUE|.
+The function may only take one argument, which is the contents of the command
+line.
+
+Every time a command needs to be previewed, Nvim executes the command. The
+command itself must do the job of making appropriate changes to the buffer
+and setting the appropriate highlights for previewing. It's possible to check
+if a command is being executed for previewing by checking if current mode name
+starts with "cp" (see |mode()|).
+
+In order for command preview to work properly, the command must follow these
+steps, in order, to implement the Nvim command preview "protocol".
+    1. Make necessary buffer changes and highlighting for the preview.
+    2. If 'inccommand' equals "split" and Nvim is in command preview mode,
+       open preview window using |vim.cmd_preview.open_win()| and put the
+       necessary text and highlighting on it. Otherwise, close preview window
+       if it's open using |vim.cmd_preview.close_win()|.
+    3. Trigger a |:redraw| of the window.
+    4. Undo all changes made to the buffer in step 1, without keeping them
+       in the undo history (see |:undo|).
+    5. Clear all highlights set in step 1.
+    6. If preview window was opened in step 2, switch back to main window to
+       ensure that if the command is executed, it gets executed on the main
+       window and not the preview window.
+
+Here is an example written in |Lua| that uses the preview flag to preview a
+command that trims trailing whitespaces from lines while following the
+aforementioned steps: >
+
```

Comment:
Oh wow, that is a huge example. I haven't taken the time to read all of it yet, but I can already say that I fear that the purpose of the example (showcasing the preview feature) might get lost in the code.

Reply:
No way around that sadly...